File: app/src/main/java/be/digitalia/fosdem/parsers/RoomStatusesParser.java
package be.digitalia.fosdem.parsers;

import android.util.JsonReader;

import java.util.HashMap;
import java.util.Map;

import be.digitalia.fosdem.model.RoomStatus;

public class RoomStatusesParser extends AbstractJsonPullParser<Map<String, RoomStatus>> {

	@Override
	protected Map<String, RoomStatus> parse(JsonReader reader) throws Exception {
		Map<String, RoomStatus> result = new HashMap<>();

		reader.beginArray();
		while (reader.hasNext()) {
			String roomName = null;
			RoomStatus roomStatus = null;

			reader.beginObject();
			while (reader.hasNext()) {
				switch (reader.nextName()) {
					case "roomname":
						roomName = reader.nextString();
						break;
					case "state":
						String stateValue = reader.nextString();
						try {
							roomStatus = RoomStatus.values()[Integer.parseInt(stateValue)];
						} catch (Exception e) {
							// Swallow and ignore that room
						}
						break;
					default:
						reader.skipValue();
				}
			}
			reader.endObject();

			if (roomName != null && roomStatus != null) {
				result.put(roomName, roomStatus);
			}
		}
		reader.endArray();

		return result;
	}
}

File: app/src/main/java/be/digitalia/fosdem/parsers/RoomStatusesParser.kt
package be.digitalia.fosdem.parsers

import be.digitalia.fosdem.model.RoomStatus
import com.squareup.moshi.JsonReader
import okio.BufferedSource

class RoomStatusesParser : Parser<Map<String, RoomStatus>> {

    override fun parse(source: BufferedSource): Map<String, RoomStatus> {
        val reader = JsonReader.of(source)
        val result = mutableMapOf<String, RoomStatus>()

        reader.beginArray()
        while (reader.hasNext()) {
            var roomName: String? = null
            var roomStatus: RoomStatus? = null

            reader.beginObject()
            while (reader.hasNext()) {
                when (reader.selectName(PROPERTIES_NAMES)) {
                    PROPERTIES_NAME_ROOM_NAME -> roomName = reader.nextString()
                    PROPERTIES_NAME_STATE -> {
                        val stateValue = reader.nextInt()
                        try {
                            roomStatus = RoomStatus.entries[stateValue]
                        } catch (e: Exception) {
                            // Swallow and ignore that room
                        }
                    }
                    else -> {
                        reader.skipName()
                        reader.skipValue()
                    }
                }
            }
            reader.endObject()

            if (roomName != null && roomStatus != null) {
                result[roomName] = roomStatus
            }
        }
        reader.endArray()

        return result
    }

    companion object {
        private val PROPERTIES_NAMES = JsonReader.Options.of(
                "roomname", "state"
        )
        private const val PROPERTIES_NAME_ROOM_NAME = 0
        private const val PROPERTIES_NAME_STATE = 1
    }
}